pupil_src/shared_modules/surface_tracker/surface.py

```diff
@@ -0,0 +1,513 @@
+        self.build_up_status = self._avg_obs_per_marker / self._required_obs_per_marker
+
+        if self.build_up_status >= 1:
+            self._finalize_def()
+
+    def _bounding_quadrangle(self, verts):
+        hull = cv2.convexHull(verts, clockwise=False)
+
+        if hull.shape[0] > 4:
+            new_hull = cv2.approxPolyDP(hull, epsilon=1, closed=True)
+            if new_hull.shape[0] >= 4:
+                hull = new_hull
+
+        if hull.shape[0] > 4:
+            curvature = abs(methods.GetAnglesPolyline(hull, closed=True))
+            most_acute_4_threshold = sorted(curvature)[3]
+            hull = hull[curvature <= most_acute_4_threshold]
+
+        # verts space is flipped in y.
+        # we need to change the order of the hull vertecies
+        hull = hull[[1, 0, 3, 2], :, :]
+
+        # now we need to roll the hull verts until we have the right orientation:
+        # verts space has its origin at the image center.
+        # adding 1 to the coordinates puts the origin at the top left.
+        distance_to_top_left = np.sqrt(
+            (hull[:, :, 0] + 1) ** 2 + (hull[:, :, 1] + 1) ** 2
+        )
+        bot_left_idx = np.argmin(distance_to_top_left) + 1
+        hull = np.roll(hull, -bot_left_idx, axis=0)
+        return hull
+
+    def _get_trans_to_norm(self, verts):
+        norm_corners = np.array([(0, 0), (1, 0), (1, 1), (0, 1)], dtype=np.float32)
+        return cv2.getPerspectiveTransform(verts, norm_corners)
+
+    def _denormalize(self, points, camera_model):
+        if len(points.shape) == 1:
+            points[1] = 1 - points[1]
+        else:
+            points[:, 1] = 1 - points[:, 1]
+        points *= camera_model.resolution
+        return points
+
+    def _finalize_def(self):
+        """
+        - prune markers that have been visible in less than x percent of frames.
+        - of those markers select a good subset of uv coords and compute mean.
+        - this mean value will be used from now on to estable surface transform
+        """
+        persistent_markers = {}
+        persistent_markers_dist = {}
+        for (k, m), m_dist in zip(
+            self.reg_markers_undist.items(), self.reg_markers_dist.values()
+        ):
+            if len(m.observations) > self._required_obs_per_marker * .5:
+                persistent_markers[k] = m
+                persistent_markers_dist[k] = m_dist
+        self.reg_markers_undist = persistent_markers
+        self.reg_markers_dist = persistent_markers_dist
+
+    def move_corner(self, corner_idx, new_pos, camera_model):
+        """Update surface definition by moving one of the corners to a new position.
+
+        Args:
+            corner_idx: Identifier of the corner to be moved. The order of corners is
+            `[(0, 0), (1, 0), (1, 1), (0, 1)]`
+            new_pos: The updated position of the corner in image pixel coordinates.
+            camera_model: Camera Model object.
+
+        """
+        # Markers undistorted
```

Comment:
The parts for undistorted and distorted markers look quite identical, can you maybe extract an method with some appropriate parameters?